### ISSUE TITLE
Use Turborepo for npm script optimizations

### ISSUE THREAD

**Issue by dotNomad** [posit-dev/publisher#1292]
On hold until we actually have a monorepo setup with NPM workspaces: #1291

---

[Turborepo](https://turbo.build/repo) adds npm script optimizations by caching them and running them in parallel. It is what [quarto](https://github.com/quarto-dev/quarto) uses for its monorepo.

**Comment by dotNomad** [comment]
If we do pick this up Turborepo does have a guide for [Using AI with Turborepo](https://turborepo.dev/docs/guides/ai) including a skill.

**Comment by dotNomad** [comment]
Unblocked by #1291 (npm workspaces) — writing up what we learned during the migration so the Turborepo work can land without re-discovering the same gotchas.

## Workspace layout that shipped

Root `package.json`:

```json
"workspaces": [
  "packages/*",
  "extensions/vscode",
  "extensions/vscode/webviews/homeView",
  "test/connect-api-contracts",
  "test/extension-contract-tests"
]
```

`test/e2e` deliberately excluded — its Cypress install weight and `postinstall` op-cli hook fit poorly in the workspace install flow.

All `"file:../../packages/..."` refs were swapped for `"*"` so npm resolves them to the local workspace. Single root `package-lock.json` replaces six sub-lockfiles.

## Aligned versions (clean inputs for turbo cache hashing)

Every workspace declares:

- `typescript: ^5.9.3`
- `vitest: ^4.1.4`
- `@types/node: ^22.0.0`

DevDeps are **kept per-package** (not hoisted to root). We considered hoisting but reverted after reviewing Turborepo examples — per-package declarations let turbo's cache keys invalidate only for affected packages when a shared tool bumps, instead of flipping every package's hash.

## vsce × workspaces gotcha (biggest surprise)

Once npm hoists to the workspace root, `vsce package` fails with `ERROR invalid relative path: extension/../../node_modules/...` because the archive root is `extensions/vscode/` and files under `../../node_modules` can't be expressed as valid archive paths. Known vsce issue: https://github.com/microsoft/vscode-vsce/issues/580.

The community-consensus fix (documented across several comments in that issue) is:

1. **Opt out of vsce's dependency walk** — we used `"vsce": { "dependencies": false }` in `extensions/vscode/package.json` (equivalent to `--no-dependencies` CLI flag, but declarative).
2. **Make the extension fully self-contained** — esbuild already bundles every runtime JS dep into `dist/extension.js`. The only asset that needed explicit handling was `@vscode/codicons/dist/{codicon.css, codicon.ttf}`, which the webview loads at runtime via `extensionUri`. We added a `copy-codicons.mjs` build script (uses `require.resolve('@vscode/codicons/package.json')` so it's hoist-agnostic) that drops the assets in `extensions/vscode/dist/codicons/` during every `esbuild-base` run, and updated `src/views/homeView.ts` to load from that path.

Net result: the `.vsix` no longer includes any `node_modules/` files. Cleaner than before.

### Secondary `vsce ls` quirk

`vsce ls` does **not** honor the `"vsce": { "dependencies": false }` in `package.json`. It prints ~3000 files with `../../` leaks. Only `vsce package` respects the setting. For diagnostics, pass `--no-dependencies` explicitly:

```
npx @vscode/vsce ls --no-dependencies
```

## CI workflow changes

Five workflows needed direct edits because they either (a) cached a specific sub-`package-lock.json` path that no longer exists, or (b) called `npm ci` in multiple sub-packages:

- `connect-contract-tests.yaml` — cache path → root lockfile, collapsed two `npm ci` steps into one
- `extension-contract-tests.yaml` — same, plus `working-directory: ${{ github.workspace }}` to escape the job-level `defaults.run.working-directory`
- `home-view-unit-test.yaml` — same pattern
- `prepare-release.yaml` — `npm install --prefix extensions/vscode` → `npm install`; `git add extensions/vscode/package-lock.json` → root `package-lock.json`
- `nightly-prerelease.yaml` — same `git add` path fix, plus don't `cd` into `extensions/vscode` before `npm install`

Workflows using `cache-dependency-path: "**/package-lock.json"` just hash fewer files and continue to work — no edits needed.

## Proposed `turbo.json` shape

```jsonc
{
  "$schema": "https://turbo.build/schema.json",
  "tasks": {
    "lint": {
      "inputs": [
        "src/**",
        "*.config.{js,ts,mjs}",
        "tsconfig*.json",
        "package.json",
      ],
    },
    "build": {
      "dependsOn": ["^build"],
      "outputs": ["dist/**"],
    },
    "test": {
      "dependsOn": ["build"],
      "inputs": ["src/**", "test/**", "tsconfig*.json", "package.json"],
    },
    "package": {
      // extension only; produces the .vsix
      "dependsOn": ["build"],
      "outputs": ["../../dist/*.vsix"],
    },
  },
}
```

Shared packages (`packages/connect-api`, `packages/connect-cloud-api`) get `build` as a no-op or light typecheck — they're consumed as source by esbuild and don't need a compiled output. The homeView webview gets its own `build` node (vue-tsc + vite) that the extension `build` depends on.

## Scope sequencing — keep the first turbo PR small

Two decisions that are tempting to bundle but are cleaner as follow-ups after the initial turbo integration lands:

1. **`just` → direct `turbo` / `npm run`.** The long-term direction is to remove `just` in favor of the native entrypoints. For the first turbo PR, the simplest path is to have `just` recipes delegate to `turbo run …` so nothing changes for contributors mid-flight. A follow-up can then retire `just` and update docs to call `turbo` / `npm run` directly.

2. **Split the homeView webview into its own turbo pipeline node.** The webview's build (`vue-tsc && vite build`) is the most expensive single step in the repo and is independent of the extension's esbuild step, so it wants its own cache entry eventually. But for the first PR, modelling the extension + webview as one pipeline node keeps `turbo.json` and the task graph simple. A follow-up can promote the webview to a first-class node once the substrate is green.

## What else to watch for

- **Don't hoist devDeps to root** — we tried, reverted. Breaks turbo cache locality.
- **`copy-codicons` is wired into `esbuild-base`** so it runs on every dev build. If turbo caches esbuild output, codicons need to be in `outputs` too (`"dist/codicons/**"`).
- **`.vscodeignore`** was simplified but still has a `node_modules` entry. Harmless with `vsce.dependencies: false`, but don't re-add negations for `node_modules/` paths without re-reading the vsce behavior above.

PR for #1291: https://github.com/posit-dev/publisher/pull/4056

Unblocks #1292 for the next implementer.
